src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java

```diff
@@ -348,7 +348,7 @@ public Next receive(Object o) {
             for (BodyExecutionCallback c : callbacks) {
                 c.onFailure(sc, t);
             }
-
+            thread.popContextVariables();
```

Comment:
I think the earlier call to `setOutcome` here and in SuccessAdapter keep us from needing to worry about this being called concurrently or more than once.